## Genomic language model pretraining with the HealthOmics sequence store

In this repo we show how to use genomic language models on the AWS cloud.

A genomic language model is a type of machine learning model, typically based on architectures like transformers or recurrent neural networks (RNNs), that is trained to understand and generate sequences of DNA, RNA, or other biological sequences. Just as language models are trained on human text to predict and generate natural language, genomic language models are trained on nucleotide sequences (like those composed of the bases A, T, C, and G) to capture the underlying patterns and structures in genomic data.

We show here, how to work with four genomic language models---[HyenaDNA](https://hazyresearch.stanford.edu/blog/2023-06-29-hyena-dna), [Evo](https://arcinstitute.org/news/blog/evo), [Evo 2](https://github.com/ArcInstitute/evo2/blob/main/README.md), and [Caduceus](https://arxiv.org/pdf/2403.03234)---and a sc-RNAseq foundational model, [Geneformer](https://www.nature.com/articles/s41586-023-06139-9).

---
### HyenaDNA

See [here](hyena-DNA/README.md) our HyenaDNA project.

---
### Evo

See [here](evo-model/README.md) our Evo project.

---
### Evo 2

See [here](evo2-model/README.md) our Evo 2 project.

---
### Geneformer

See [here](geneformer/README.md) our Geneformer project.

---
### Caduceus

See [here](caduceus/README.md) our Caduceus project.

---
## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.


## License

This library is licensed under the MIT-0 License. See the [LICENSE](LICENSE) file.
